File: src/LeetCode/L76/Main.java
package LeetCode.L76;

import java.util.HashMap;
import java.util.Map;

/**
 * 给你一个字符串 s 、一个字符串 t 。返回 s 中涵盖 t 所有字符的最小子串。如果 s 中不存在涵盖 t 所有字符的子串，则返回空字符串 "" 。
 *
 *
 *
 * 注意：
 *
 * 对于 t 中重复字符，我们寻找的子字符串中该字符数量必须不少于 t 中该字符数量。
 * 如果 s 中存在这样的子串，我们保证它是唯一的答案。
 *
 *
 * 示例 1：
 *
 * 输入：s = "ADOBECODEBANC", t = "ABC"
 * 输出："BANC"
 * 解释：最小覆盖子串 "BANC" 包含来自字符串 t 的 'A'、'B' 和 'C'。
 * 示例 2：
 *
 * 输入：s = "a", t = "a"
 * 输出："a"
 * 解释：整个字符串 s 是最小覆盖子串。
 * 示例 3:
 *
 * 输入: s = "a", t = "aa"
 * 输出: ""
 * 解释: t 中两个字符 'a' 均应包含在 s 的子串中，
 * 因此没有符合条件的子字符串，返回空字符串。
 *
 *
 * 提示：
 *
 * m == s.length
 * n == t.length
 * 1 <= m, n <= 10^5
 * s 和 t 由英文字母组成
 *
 *
 * 进阶：你能设计一个在 o(m+n) 时间内解决此问题的算法吗？
 */
public class Main {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println("s.minWindow(\"ADOBECODEBANC\", \"ABC\") = " + s.minWindow("ADOBECODEBANC", "ABC"));
    }
}

class Solution {
    private final Map<Character, Integer> tcnts = new HashMap<>();
    private final Map<Character, Integer> cnt = new HashMap<>();

    public String minWindow(String s, String t) {
        for (char c : t.toCharArray()) {
            Integer cnt = tcnts.get(c);
            if (cnt == null){
                cnt = 1;
            } else {
                cnt++;
            }
            tcnts.put(c, cnt);
        }
        int l = 0, r = -1;
        int len = Integer.MAX_VALUE, ansL = -1, ansR = -1;
        int sLen = s.length();
        while (r < sLen) {
            r++;
            if (r < sLen && tcnts.containsKey(s.charAt(r))) {
                cnt.put(s.charAt(r), cnt.getOrDefault(s.charAt(r), 0) + 1);
            }
            while (check() && l <= r) {
                if (r - l + 1 < len) {
                    len = r - l + 1;
                    ansL = l;
                    ansR = l + len;
                }
                if (tcnts.containsKey(s.charAt(l))) {
                    cnt.put(s.charAt(l), cnt.getOrDefault(s.charAt(l), 0) - 1);
                }
                ++l;
            }
        }
        return ansL == -1 ? "" : s.substring(ansL, ansR);
    }

    private boolean check() {
        for (Map.Entry<Character, Integer> characterIntegerEntry : tcnts.entrySet()) {
            Character key = (Character) ((Map.Entry<?, ?>) characterIntegerEntry).getKey();
            Integer val = (Integer) ((Map.Entry<?, ?>) characterIntegerEntry).getValue();
            if (cnt.getOrDefault(key, 0) < val) {
                return false;
            }
        }
        return true;
    }
}

File: src/LeetCode/L76/main.go
package main

import "fmt"

/*
*
给你一个字符串 s 、一个字符串 t 。返回 s 中涵盖 t 所有字符的最小子串。如果 s 中不存在涵盖 t 所有字符的子串，则返回空字符串 "" 。

注意：

对于 t 中重复字符，我们寻找的子字符串中该字符数量必须不少于 t 中该字符数量。
如果 s 中存在这样的子串，我们保证它是唯一的答案。

示例 1：

输入：s = "ADOBECODEBANC", t = "ABC"
输出："BANC"
解释：最小覆盖子串 "BANC" 包含来自字符串 t 的 'A'、'B' 和 'C'。
示例 2：

输入：s = "a", t = "a"
输出："a"
解释：整个字符串 s 是最小覆盖子串。
示例 3:

输入: s = "a", t = "aa"
输出: ""
解释: t 中两个字符 'a' 均应包含在 s 的子串中，
因此没有符合条件的子字符串，返回空字符串。

提示：

- m == s.length
- n == t.length
- 1 <= m, n <= 10^5
- s 和 t 由英文字母组成
*/
func main() {
	fmt.Println("minWindow(\"ADOBECODEBANC\", \"ABC\") = ", minWindow("ADOBECODEBANC", "ABC"))
}

func minWindow(s string, t string) string {
	m, n := len(s), len(t)
	scnt, tcnt := make(map[uint8]int, 0), make(map[uint8]int, 0)
	for i := 0; i < n; i++ {
		tcnt[t[i]]++
	}
	check := func() bool {
		for k, v := range tcnt {
			if scnt[k] < v {
				return false
			}
		}
		return true
	}
	var (
		ansl   = -1
		ansr   = -1
		anslen = m + 1
	)
	for l, r := 0, 0; r < m; r++ {
		if r < m && tcnt[s[r]] > 0 {
			scnt[s[r]]++
		}
		for check() && l <= r {
			if r-l+1 < anslen {
				anslen = r - l + 1
				ansl, ansr = l, l+anslen
			}
			if _, ok := tcnt[s[l]]; ok {
				scnt[s[l]] -= 1
			}
			l++
		}
	}
	if anslen > m {
		return ""
	}
	return s[ansl:ansr]
}
